File: alds/queue/src/queue.rb
# ALDS1_3_B: Queue

# A simple queue.
class Queue
  def initialize(capacity)
    @queue = Array.new capacity
    @head = @num = 0
  end

  def enq(item)
    raise 'Queue is full' if full?
    @num += 1
    tail = (@head + @num - 1) % @queue.size
    @queue[tail] = item
  end

  def deq
    raise 'Queue is empty' if empty?
    item = @queue[@head]
    @queue[@head] = nil
    @head = (@head + 1) % @queue.size
    @num -= 1
    item
  end

  def empty?
    @num == 0
  end

  def full?
    @num == @queue.size
  end
end

# A processor.
class CPU
  MAX_N_PROCESSES = 100000
  Process = Struct.new :name, :time

  def initialize(quantum)
    @quantum = quantum
  end

  def load_ps(n_ps)
    ps = Queue.new MAX_N_PROCESSES
    n_ps.times do
      input = gets.chomp.split
      process = Process.new input[0], input[1].to_i
      ps.enq process
    end
    ps
  end

  def run(ps)
    elapsed_time = 0

    until ps.empty?
      process = ps.deq
      if process.time <= @quantum
        elapsed_time += process.time
        puts "#{process.name} #{elapsed_time}"
      else
        elapsed_time += @quantum
        process.time -= @quantum
        ps.enq process
      end
    end
  end
end

# Run processes.
n_ps, quantum = gets.split.map &:to_i
cpu = CPU.new(quantum)
cpu.run cpu.load_ps(n_ps)

File: alds/queue/src/queue.cpp
// 配列によるQueueの実装
//
// Queueの先頭を指し示すポインタheadと末尾の次の要素(次にenqueueされる値が入る場所)を
// 指し示すポインタtailを用意する。要素が取り出されたら head++, 追加されたら tail++。
// head == tail の時はQueueが空ということ。ただしこれでは取り出された後の配列の領域が
// 無駄になるため、配列をリングバッファとして扱うようにする。tail が配列末尾に達したら0に
// リセットし、続く要素は配列の頭から格納し直していく。head - tail == 1 になったら(headの
// 2つ前にtailが来たら)Queueは満杯とする(tailをheadに追いつかせてしまうと、空なのか満杯なのか
// 区別できないため)。

// 与えられるプロセスをQueueに格納
// 順番に処理していき、クォンタム以内に終わらないものは再度Queueへ。
// (あるいは、終わらなかったもののみQueueへ入れてもいいのかも)。
// 後は全てのプロセスが終わるまでQueueをループ。
// 終わったプロセスを終了タイムとともに順に結果配列へ格納しておき、
// 最後にそれを出力する。
//
// 経過時間　処理開始から一つのプロセスを処理するごとにプラスされる。
// そのプロセスがクォンタム以内に終わればそのプロセスの処理時間を、
// 終わらなければクォンタム分を追加する。

#include <iostream>
#include <cstdlib>
#include <queue>
using namespace std;

struct Process {
    string name;
    int time;
};

/* ALDS1_3_B: Queue */
int main() {
    int n, quantum;
    cin >> n >> quantum;

    queue<Process*> pqueue;
    Process *ps;

    for (int i = 0; i < n; i++) {
        ps = new Process();
        cin >> ps->name >> ps->time;
        pqueue.push(ps);
    }

    int elapsed = 0;
    while(! pqueue.empty() ) {
        ps = pqueue.front();
        pqueue.pop();

        if (quantum < ps->time) {
            elapsed += quantum;
            ps->time -= quantum;
            pqueue.push(ps);
        } else {
            elapsed += ps->time;
            cout << ps->name << " " << elapsed << endl;
        }
    }
    free(ps);

    return 0;
}
